src/renderer/components/settings-execution.tsx

```diff
@@ -59,36 +101,201 @@ export class ExecutionSettings extends React.Component<ExecutionSettingsProps> {
+    const { executionFlags, environmentVariables } = this.state;
+    const { appState } = this.props;
+
+    if (type === SettingItemType.Flag) {
+      const flags = Object.values(executionFlags);
+      appState.executionFlags = flags.filter((f) => f !== '');
+    } else {
+      const vars = Object.values(environmentVariables);
+      appState.environmentVariables = vars.filter((v) => v !== '');
+    }
   }
 
   /**
-   * Handles a change in the environment variables run with the Electron executable.
+   * Handles a change in the execution flags or environment variables
+   * run with the Electron executable.
    *
    * @param {React.ChangeEvent<HTMLInputElement>} event
+   * @param {SettingItemType} type
    */
-  public handleEnvironmentVariableChange(
-    event: React.FormEvent<HTMLInputElement>,
+  public handleSettingsItemChange(
+    event: React.ChangeEvent<HTMLInputElement>,
+    type: SettingItemType,
   ) {
+    const { name, value } = event.currentTarget;
+
+    if (type === SettingItemType.Flag) {
+      this.setState((prevState) => ({
+        executionFlags: {
+          ...prevState.executionFlags,
+          [name]: value,
+        },
+      }));
+    } else {
+      this.setState((prevState) => ({
+        environmentVariables: {
+          ...prevState.environmentVariables,
+          [name]: value,
+        },
+      }));
+    }
+  }
+
+  /**
+   * Adds a new settings item input field.
+   *
+   * @param {SettingItemType} type
+   */
+  private addNewSettingsItem(type: SettingItemType) {
+    if (type === SettingItemType.Flag) {
+      const flagsArray = Object.entries(this.state.executionFlags);
+
+      this.setState((prevState) => ({
+        executionFlags: {
+          ...prevState.executionFlags,
+          [`flag-${flagsArray.length}`]: '',
+        },
+      }));
+    } else {
+      const varsArray = Object.entries(this.state.environmentVariables);
+
+      this.setState((prevState) => ({
+        environmentVariables: {
+          ...prevState.environmentVariables,
+          [`var-${varsArray.length}`]: '',
+        },
+      }));
+    }
+  }
+
+  /**
+   * Handle a change to the package manager used to install modules when running
+   * Fiddles;
+   *
+   * @param {React.FormEvent<HTMLInputElement>} event
+   */
+  private handlePMChange = (event: React.FormEvent<HTMLInputElement>) => {
+    const { appState } = this.props;
     const { value } = event.currentTarget;
 
-    const variables = value.split('|');
-    this.props.appState.environmentVariables = variables;
+    appState.packageManager = value as IPackageManager;
+  };
+
+  public renderDeleteItem(idx: string, type: SettingItemType): JSX.Element {
+    const clickFn = () => {
+      if (type === SettingItemType.Flag) {
+        const flags = this.state.executionFlags;
+        if (Object.keys(flags).length === 1) {
+          flags[idx] = '';
+        } else {
+          delete flags[idx];
+        }
```

Comment:
I don't understand why there's special logic for not deleting the last property? Do we need to avoid an empty object {} for some reason?

I'm not arguing pro or con on this; literally asking because I don't see why the special case is needed

Reply:
@ckerr if a user is removing the only item, i don't want to not show any input fields - open to other ways to do this though! instead  i could disable the remove button for example?

Reply:
That's a good idea; I like that better.